In [1]:
#Importing the packages we'll be leveraging in this project
import pandas as pd
import seaborn as sns
import numpy as np
import matplotlib.pyplot as plt

In [35]:
# load in the dataset into a pandas dataframe, print statistics
url = 'https://github.com/HamoyeHQ/HDSC-Introduction-to-Python-for-machine-learning/files/7768140/FoodBalanceSheets_E_Africa_NOFLAG.csv'
fb = pd.read_csv(url, encoding= 'latin-1')
fb.head()

Area Code     Area  Item Code         Item  Element Code  \
0          4  Algeria       2501   Population           511   
1          4  Algeria       2501   Population          5301   
2          4  Algeria       2901  Grand Total           664   
3          4  Algeria       2901  Grand Total           674   
4          4  Algeria       2901  Grand Total           684   

                                  Element             Unit     Y2014  \
0           Total Population - Both sexes     1000 persons  38924.00   
1                Domestic supply quantity      1000 tonnes      0.00   
2           Food supply (kcal/capita/day)  kcal/capita/day   3377.00   
3  Protein supply quantity (g/capita/day)     g/capita/day     94.90   
4      Fat supply quantity (g/capita/day)     g/capita/day     80.06   

      Y2015     Y2016     Y2017     Y2018  
0  39728.00  40551.00  41389.00  42228.00  
1      0.00      0.00      0.00      0.00  
2   3379.00   3372.00   3341.00   3322.00  
3     94.35     94.72     92.82     91.83  
4     79.36     77.40     80.19     77.28

In [3]:
# high-level overview of data shape and composition
print(fb.shape)
print(fb.dtypes)

(60943, 12)
Area Code         int64
Area             object
Item Code         int64
Item             object
Element Code      int64
Element          object
Unit             object
Y2014           float64
Y2015           float64
Y2016           float64
Y2017           float64
Y2018           float64
dtype: object


In [4]:
print(fb.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60943 entries, 0 to 60942
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Area Code     60943 non-null  int64  
 1   Area          60943 non-null  object 
 2   Item Code     60943 non-null  int64  
 3   Item          60943 non-null  object 
 4   Element Code  60943 non-null  int64  
 5   Element       60943 non-null  object 
 6   Unit          60943 non-null  object 
 7   Y2014         59354 non-null  float64
 8   Y2015         59395 non-null  float64
 9   Y2016         59408 non-null  float64
 10  Y2017         59437 non-null  float64
 11  Y2018         59507 non-null  float64
dtypes: float64(5), int64(3), object(4)
memory usage: 5.6+ MB
None


In [5]:
# the sum of null values in each colu
fb.isnull().sum()

Area Code          0
Area               0
Item Code          0
Item               0
Element Code       0
Element            0
Unit               0
Y2014           1589
Y2015           1548
Y2016           1535
Y2017           1506
Y2018           1436
dtype: int64

In [6]:
fb.dropna().count()

Area Code       57784
Area            57784
Item Code       57784
Item            57784
Element Code    57784
Element         57784
Unit            57784
Y2014           57784
Y2015           57784
Y2016           57784
Y2017           57784
Y2018           57784
dtype: int64

In [7]:
fb['Element'].unique()

array(['Total Population - Both sexes', 'Domestic supply quantity',
       'Food supply (kcal/capita/day)',
       'Protein supply quantity (g/capita/day)',
       'Fat supply quantity (g/capita/day)', 'Production',
       'Import Quantity', 'Stock Variation', 'Export Quantity', 'Feed',
       'Seed', 'Losses', 'Processing', 'Other uses (non-food)',
       'Residuals', 'Food', 'Food supply quantity (kg/capita/yr)',
       'Tourist consumption'], dtype=object)

### 1. What is the mean and standard deviation across the whole dataset for the year 2017 to 2 decimal places?

In [8]:
fb['Y2017'].describe()

count     59437.000000
mean        140.917765
std        1671.862359
min       -1582.000000
25%           0.000000
50%           0.100000
75%           9.000000
max      190873.000000
Name: Y2017, dtype: float64

### 2. Given the following python code, what would the output of the code give?

In [9]:
my_tuppy = (1,2,5,8)

In [10]:
my_tuppy[2] = 6

TypeError: 'tuple' object does not support item assignment

### 3. Select columns ‘Y2017’ and ‘Area’, Perform a groupby operation on ‘Area’.  Which of these Areas had the 7th lowest sum in 2017?

In [11]:
fb[['Y2017', 'Area']].groupby('Area').sum().sort_values(by='Y2017', descending=False)

TypeError: sort_values() got an unexpected keyword argument 'descending'

### 4. 
#### Consider the following list of tuples: y = [(2, 4), (7, 8), (1, 5, 9)]
#### How would you assign element 8 from the list to a variable x? 

In [ ]:
y = [(2,4), (7,8), (1,5,9)]
x = y[1][-1]
print(x)

In [ ]:
y = [(2,4), (7,8), (1,5,9)]
x = y[1][1]
print(x)

###  5. 
What would be the output for?
 S = [['him', 'sell'], [90, 28, 43]]
 S[0][1][1]



In [13]:
S = [['him', 'sell'], [90, 28, 43]]

In [14]:
S[0][1][1]

'e'

### 6. Which of these python data structures is unorderly?

Set and Dictionary

### 7. Select columns ‘Y2017’ and ‘Area’, Perform a groupby operation on ‘Area’.  Which of these Areas had the highest sum in 2017?

In [12]:
fb[['Y2017', 'Area']].groupby('Area').sum().sort_values(by='Y2017', ascending=False)

Y2017
Area                                   
Nigeria                      1483268.23
Egypt                         866379.92
South Africa                  517590.54
Ethiopia                      448683.76
Morocco                       388495.36
Ghana                         337599.06
Algeria                       325644.27
United Republic of Tanzania   322616.85
Kenya                         264660.66
Sudan                         239931.92
Cameroon                      232030.43
Angola                        229159.57
Côte d'Ivoire                 224599.01
Uganda                        213950.38
Malawi                        181098.71
Mozambique                    161407.98
Mauritania                    156665.46
Mali                          149928.33
Madagascar                    131197.73
Niger                         126707.58
Benin                         124771.22
Tunisia                       124167.20
Zambia                        103223.77
Burkina Faso                  101855.07
Guinea                         98138.87
Senegal                        95681.15
Zimbabwe                       75919.34
Rwanda                         73663.69
Chad                           71594.68
Sierra Leone                   55311.33
Eswatini                       54343.33
Mauritius                      51114.83
Togo                           49841.88
Congo                          41181.68
Central African Republic       29937.00
Namibia                        29874.89
Liberia                        29342.20
Gabon                          27979.64
Gambia                         23154.18
Djibouti                       22729.91
Botswana                       22101.30
Lesotho                        21267.96
Guinea-Bissau                  19102.77
Cabo Verde                     14650.74
Sao Tome and Principe          12662.63
Seychelles                       442.34
Comoros                           59.84
Sudan (former)                     0.00
Ethiopia PDR                       0.00

In [ ]:
Ans = Nigeria

### 8. Which of the following dataframe methods can be used to access elements across rows and columns?

In [ ]:
C and D

### 9. Which year had the least correlation with ‘Element Code’?

In [16]:
fb.corr()['Element Code'].sort_values(ascending=False)

Element Code    1.000000
Y2014           0.024457
Y2018           0.024279
Y2017           0.024254
Y2015           0.023889
Y2016           0.023444
Area Code      -0.000209
Item Code      -0.024683
Name: Element Code, dtype: float64

In [17]:
Answer = 2016

### 10. If you have the following list

In [21]:
lst = [[35, 'Portugal', 94], [33, 'Argentina', 93], [30, 'Brazil', 92]]
col = ['Age', 'Nationality', 'Overall']

In [22]:
pd.DataFrame(lst, columns=col, index=[1,2,3])

Age Nationality  Overall
1   35    Portugal       94
2   33   Argentina       93
3   30      Brazil       92

In [23]:
pd.DataFrame(lst, columns=['Age', 'Nationality', 'Overall'], index=[1,2,3])

Age Nationality  Overall
1   35    Portugal       94
2   33   Argentina       93
3   30      Brazil       92

### 11. What is the total Protein supply quantity in Madagascar in 2015?

In [24]:
fb[fb['Area'] == 'Madagascar'].groupby('Element').sum()

Area Code  Item Code  Element Code  \
Element                                                                      
Domestic supply quantity                    15351     318888        630819   
Export Quantity                             13416     279445        614744   
Fat supply quantity (g/capita/day)          14319     299405         75924   
Feed                                         4515      94030        193235   
Food                                        14190     295817        565620   
Food supply (kcal/capita/day)               14319     299405         73704   
Food supply quantity (kg/capita/yr)         14190     295817         70950   
Import Quantity                             14319     297945        622821   
Losses                                       7095     148102        281765   
Other uses (non-food)                        6579     138246        262854   
Processing                                   9804     201929        389956   
Production                                  14190     295071        606210   
Protein supply quantity (g/capita/day)      14319     299405         74814   
Residuals                                   13545     280011        542850   
Seed                                         2709      56067        116067   
Stock Variation                             12642     262174        497056   
Total Population - Both sexes                 129       2501           511   

                                           Y2014     Y2015     Y2016  \
Element                                                                
Domestic supply quantity                31578.76  31214.98  31732.08   
Export Quantity                           425.48    494.75    535.06   
Fat supply quantity (g/capita/day)         88.72     91.85     95.33   
Feed                                     2178.76   2070.22   2121.40   
Food                                    21405.42  21120.65  21435.61   
Food supply (kcal/capita/day)            7945.00   7685.00   7669.00   
Food supply quantity (kg/capita/yr)       907.59    871.59    861.17   
Import Quantity                          2010.71   1721.80   1901.47   
Losses                                   2252.00   2176.00   2215.00   
Other uses (non-food)                    4348.58   3957.12   4114.08   
Processing                               1262.00   1803.00   1770.00   
Production                              30337.51  29482.89  29889.65   
Protein supply quantity (g/capita/day)    177.79    173.05    173.40   
Residuals                                -603.00   -616.00   -642.00   
Seed                                      728.00    699.00    720.00   
Stock Variation                           352.00   -505.00   -475.00   
Total Population - Both sexes           23590.00  24234.00  24894.00   

                                           Y2017     Y2018  
Element                                                     
Domestic supply quantity                31927.87  32596.87  
Export Quantity                           566.81    535.81  
Fat supply quantity (g/capita/day)        101.03     96.43  
Feed                                     2059.60   2133.60  
Food                                    21676.21  22116.21  
Food supply (kcal/capita/day)            7768.00   7748.00  
Food supply quantity (kg/capita/yr)       847.75    842.17  
Import Quantity                          3139.79   3214.79  
Losses                                   2153.00   2253.00  
Other uses (non-food)                    4170.06   4221.06  
Processing                               1826.00   1819.00  
Production                              29267.86  30215.86  
Protein supply quantity (g/capita/day)    175.75    176.64  
Residuals                                -657.00   -676.00  
Seed                                      695.00    728.00  
Stock Variation                           -91.00    293.00  
Total Population - Both sexes           25571.00  26262.00

### 12. What is the total number of unique countries in the dataset?

In [25]:
fb['Area'].nunique()

49

### 13. Perform a groupby operation on ‘Element’.  What year has the highest sum of Stock Variation?

In [27]:
stock_2014 = fb.groupby('Element').sum().loc['Stock Variation']['Y2018']
print(f'The sum of the stock variation in 2014 is {round(stock_2014, 2)}')

The sum of the stock variation in 2014 is 20577.91



### 14. Perform a groupby operation on ‘Element’.  What is the total number of the sum of Processing in 2017?

In [28]:
fb[fb['Element'] == 'Processing'].groupby('Element').sum()

Area Code  Item Code  Element Code     Y2014     Y2015     Y2016  \
Element                                                                        
Processing     271940    5350416      10313310  282923.0  287929.0  280631.0   

               Y2017     Y2018  
Element                         
Processing  292836.0  308429.0


### 15. A pandas Dataframe with dimensions (100,3) has how many features and observations?

3 features, 100 observations

### 16. Answer the following questions based on the African food production dataset provided by the FAO website already provided

### What is the total sum of Wine produced in 2015 and 2018 respectively?

#### Hint:

### Perform a groupby sum aggregation on ‘Item’

In [30]:
fb[fb['Item'] == 'Wine'].groupby('Item').sum()

Area Code  Item Code  Element Code    Y2014    Y2015    Y2016    Y2017  \
Item                                                                           
Wine      66206    1319535       1818328  4497.36  4251.81  3872.09  4178.02   

        Y2018  
Item           
Wine  4039.32

Ans = 4251.81 and 4039.32

### 17. How would you check for the number of rows and columns in a pandas DataFrame named df?

df.shape


### 18. Which of the following is a python inbuilt module?

Math

In [ ]:
19. Given the following numpy array 

array  = ([[94, 89, 63],

             [93, 92, 48],

             [92, 94, 56]])

How would you select  the elements in bold and italics from the array?

Options
array[ : 2, 1 : ] 

array[ : 1, 1 : ]

array[ 1 : , : ]

array[ : 1, 1 :] 

array[ : 2, 0 : ]

### 20. What is the total number and percentage of missing data in 2014 to 3 decimal places?

In [34]:
total = fb.isnull().sum().sort_values(ascending=False)
percent1 = fb.isnull().sum()/fb.isnull().count()*100
percent2 = (round(percent1, 3)).sort_values(ascending=False)
missing_data = pd.concat([total, percent2], axis=1, keys=['Total', '%'])
missing_data.head()

Total      %
Y2014   1589  2.607
Y2015   1548  2.540
Y2016   1535  2.519
Y2017   1506  2.471
Y2018   1436  2.356

Ans = 1589 and 2.607%